### DISCUSSION TITLE
Easily showing version

### DISCUSSION THREAD

**Discussion by reagle** [timbrel/GitSavvy#1677] [Q&A] [answered]
Is there any easy way to show and navigate back/forth from a snapshot (complete files, not just the diff) in the history without having to do checkouts?

**Comment by kaste** [comment, marked as answer]
This possible for files, not complete trees.  I actually have a branch to make this better because I use it quiet often.  

You *could* use `git: show current file at commit` from the Command Palette but I usually use `[o]` when in a hunk in a commit/patch view.  

For example:  in the repo history ("graph") I see an interesting commit, `[o]` to open the commit and then again `[o]` to open the complete file at *that* commit.  Or even more so:  I select something in a normal buffer, then `Line History` which also shows patches and then `[o]` to open the complete file at that commit again. 

In these views there is typically `o` and the uppercase `O` bound.  I may flip both in the future as one navigation is to the same position as the patch shows in the checked out working dir ("live" version of that file) and one is to show the historical revision of that file.  

There is a plan to do this for complete trees actually (basically using temporary worktrees).  I personally do a lot of research, revision digging and reading while coding so I probably would like to have that feature.

**Reply by reagle** [reply]
Thank you, very helpful. When in "GRAPH" view and I use `m` to show the "detail panel", is there a keyboard shortcut that allows me to move the cursor to the "detail panel" (to scroll it) and back to "GRAPH" to continue perusing the history.

**Reply by kaste** [reply]
Oh, I actually have that in a branch.  I think I'm undecided of how I should roll out this.  I personally have it on `[tab]` right now.  Maybe `[M]` would be good.

**Comment by reagle** [comment]
My `[tab]` goes to the next view (e.g., GRAPH to STATUS).

**Reply by kaste** [reply]
Yes sure.  I have that overwritten.  At the moment `[tab]` switches focus back/forth between the graph and the commit/patch panel (if it is open) *on my branch*.  I don't like the view cycling via `tab` *that* much.  For example: I _always_ press `g` from within the status view to see the graph.)

I actually am unsure about what a good or best workflow would be to switch to the bottom panel and back.

**Reply by reagle** [reply]
If you don't use the cycling, how do you get to the branch or tag view? It'd be nice if there were single letter keys reserved for each view, but there'd be conflicts presently (e.g., `t` for tag for for stash. I can imagine `j` meaning "jump" and it pops up a modal that allows me to jump to the views (i.e., `j+t` for tag, `j+b` for branch, `j+g` for graph, which could also be the default/top entry). But otherwise, I use the default and I suppose `M` makes most sense.

**Reply by kaste** [reply]
I bound `ctrl+shift+s` for the status, `ctrl+shift+g` for the graph, and `ctrl+shift+c` for making a commit. 

I don't use tags and branches *that* often.  So I use the command palette for the former, and yeah `tab` to go to the branches.  Usually I push/pull or checkout within the graph or also using the command palette.  I make new branches with the command palette.  I also use a quick Terminus panel bound to another `ctrl+shift` key.

`j` is often taken in vim mode.  (`j/k` for previous/next thing).

**Reply by reagle** [reply]
Ah. You could use `g` for "go" with the default being graph, and other letters in the modal for the other views.

**Comment by kaste** [comment]
I looked at the `tab` binding again.  Switching feels a bit sluggish.  There is some debouncing or async'ness in play probably.  When it's `tab`ing I think it should feel smooth and hopefully instant.  (I remember that we have `tab`ing in the diff views to flip between the stage and unstaged area; and this is usually instant, and both sides of the tab are truly connected.)  

Then, from the default "dashboard" views, the rebase one is deprecated and I should remove it (from the default at least).  The functionality lives in the graph now.  And should be faster.  (The rebase dashboard is also the slowest to open.)  The graph "dashboard" should also not be in the list.  Ideally the graph is a long lived view because rendering the graph (from scratch) is very expensive.  But cycling reuses the view (or closes the previous view).  The graph also has a different styling.

So we have status, branch, tags which are unique and share the same UI.  But they don't go together UX wise.  For example: status and the diff around all files have a close relationship; or status and the commit message view.  

I do like `Gb`, `Gt` as goto commands.  The lowercase `g` is currently bound to open the graph in basically every view so can't use that.  Just thinking out loud here.

**Comment by kaste** [comment]
What I found out recently is that when I'm looking at a patch, 

![image](https://user-images.githubusercontent.com/8558/218163525-34f0f299-8233-4bf0-8367-4eb4d386a365.png)

and now want to see more context from the hunk the cursor is in.  Instead of `o` as discussed above, I can simply open the inline diff (which I also have bound to a key) and immediately see

![image](https://user-images.githubusercontent.com/8558/218163905-e1854327-39f6-4ca8-894d-68512a148a37.png)

So I basically see the patch with context.  (And then I can use `p`(revious) and `n`(ewer) to walk the versions.  Although this walking currently only works if the commits are reachable from the currently checked out HEAD.)  This also works from the Line History.
